python/ray/ml/preprocessors/encoder.py

```diff
@@ -88,33 +119,121 @@ class OneHotEncoder(Preprocessor):
             in ``columns``.
     """
 
-    def __init__(self, columns: List[str], limit: Optional[Dict[str, int]] = None):
+    def __init__(self, columns: List[str], *, limit: Optional[Dict[str, int]] = None):
         # TODO: add `drop` parameter.
         self.columns = columns
         self.limit = limit
 
     def _fit(self, dataset: Dataset) -> Preprocessor:
-        self.stats_ = _get_unique_value_indices(dataset, self.columns, limit=self.limit)
+        self.stats_ = _get_unique_value_indices(
+            dataset, self.columns, limit=self.limit, encode_lists=False
+        )
         return self
 
     def _transform_pandas(self, df: pd.DataFrame):
         _validate_df(df, *self.columns)
+
+        columns_to_drop = set(self.columns)
+
         # Compute new one-hot encoded columns
         for column in self.columns:
             column_values = self.stats_[f"unique_values({column})"]
+            if _is_series_composed_of_lists(df[column]):
+                df[column] = df[column].map(lambda x: tuple(x))
             for column_value in column_values:
                 df[f"{column}_{column_value}"] = (df[column] == column_value).astype(
                     int
                 )
         # Drop original unencoded columns.
-        df = df.drop(columns=self.columns)
+        df = df.drop(columns=list(columns_to_drop))
         return df
 
     def __repr__(self):
         stats = getattr(self, "stats_", None)
         return f"OneHotEncoder(columns={self.columns}, stats={stats})"
 
 
+class MultiHotEncoder(Preprocessor):
+    """Encode columns using multi-hot encoding.
+
+    A column of lists or scalars (treated as one element lists) will be
+    encoded as a column of one-hot encoded lists. This is useful for eg.
+    generating embeddings for recommender systems.
+
+    Example:
+
+    >>> import ray.data
+    >>> from ray.ml.preprocessors import MultiHotEncoder
+    >>> import pandas as pd
+    >>> mhe = MultiHotEncoder(columns=["A", "B"])
+    >>> batch = pd.DataFrame(
+    ...     {
+    ...         "A": [["warm"], [], ["hot", "warm", "cold"], ["cold", "cold"]],
```

Comment:
Love the example here!
Do you want to change ["warm"] to just "warm"?

Reply:
Should be all lists

Reply:
Then we probably should call out the assumption that no mixing of scaler and list is expected within one column right? The same goes for the OrdinalEncoder. It's easy for user to mix it up and no error is thrown.

Reply:
I updated the docstrings